File: 047/main.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function (nums) {
  if (nums == null || nums.length === 0) return []
  nums.sort((a, b) => a - b)
  let res = []
  let s = new Set()
  let arr = []
  dfs(arr, s)
  return res

  function dfs (arr, s) {
    if (arr.length === nums.length) {
      let mid = arr.slice()
      res.push(mid)
    } else {
      let pre = nums[0] - 1
      for (let i = 0; i < nums.length; i++) {
        if (pre !== nums[i] && !s.has(i)) {
          pre = nums[i]
          s.add(i)
          arr.push(nums[i])
          dfs(arr, s)
          s.delete(i)
          arr.splice(-1)
        }
      }
    }
  }
}

console.log(permuteUnique([2, 2, 1, 1]))

File: 047/main.cpp
class Solution {
public:
    vector<vector<int> > permuteUnique(vector<int>& nums) {
        vector<vector<int> > res;
        vector<int> arr;
        set<int> s;
        if (nums.size() == 0) return res;
        sort(nums.begin(), nums.end());
        dfs(res, arr, s, nums);
        return res;
    }
    
    void dfs(vector<vector<int> >& res, vector<int>& arr, set<int>& s, vector<int>& nums) {
        if (arr.size() == nums.size()) {
            res.push_back(arr);
        } else {
            int pre = nums[0] - 1;
            for (int i = 0; i < nums.size(); i++) {
                if (s.find(i) == s.end() && pre != nums[i]) {
                    pre = nums[i];
                    arr.push_back(nums[i]);
                    s.insert(i);
                    dfs(res, arr, s, nums);
                    arr.pop_back();
                    s.erase(i);
                }
            }
        }
    }
};